src/weight_windows.cpp

```diff
@@ -0,0 +1,284 @@
+      particle_type = openmc::ParticleType::photon;
+      p_ww = true;
+    }
+    xml_node particle = node.child("particle");
+    weight_params[particle_type] = read_particle_settings(particle);
+    
+  } else {
+    fatal_error("Weight window file is missing a particle section");   
+  }
+
+  // user defined source weight biasing in energy
+  if (check_for_node(node, "user_defined_biasing")) {
+    read_user_biasing(node);  
+  }
+}
+
+// read the particle specific settings 
+WWParams WeightWindow::read_particle_settings(pugi::xml_node node) 
+{ 
+  using namespace pugi;
+
+  WWParams settings;
+  
+  // 
+  if (!check_for_node(node, "parameters")) {
+    fatal_error("<parameters> section is missing from the" 
+      "weight_windows.xml file.");
+  } else {
+    pugi::xml_node parameters = node.child("parameters");
+  
+    // get the upper bound - optional - could be defined as a mesh parameter
+    settings.upper_ratio = std::stod(get_node_value(parameters, "upper_bound"));
+    if(settings.upper_ratio < 2) fatal_error("upper bound must be larger than 2");
+
+    // get the survival value
+    settings.survival_ratio = std::stod(get_node_value(parameters, "survival"));
+    if(settings.survival_ratio <= 1 
+      || settings.survival_ratio >= settings.upper_ratio) 
+        fatal_error("Survival to lower weight window ratio must bigger than 1"
+                  "and less than the upper to lower weight window ratio.");   
+     
+    // get the max split
+    settings.max_split = std::stod(get_node_value(parameters, "max_split"));
+    if(settings.max_split <= 1) fatal_error("max split must be larger than 1");
+     
+    // multiplier
+    settings.multiplier = std::stod(get_node_value(parameters, "multiplier"));
+    if(settings.multiplier <= 0) fatal_error("multiplier must be larger than 0");
+  }  
+
+  // 
+  if (!check_for_node(node, "energy")) {
+    fatal_error("<energy> section is missing from the " 
+      "weight_windows.xml file.");
+  } else {
+    settings.energy_bounds = get_node_array<double>(node, "energy");   
+  }
+
+  // read the lower weight bounds
+  if (!check_for_node(node, "lower_ww")) {
+    fatal_error("<lower_ww> section is missing from the " 
+      "weight_windows.xml file.");
+  } else {
+    settings.lower_ww = get_node_array<double>(node, "lower_ww");    
+    if (settings.lower_ww.size() != 
+      mesh_->n_bins()*(settings.energy_bounds.size()-1)) {
+         fatal_error("not enough entries in the lower_ww section");
+      }
+  }
+
+  // read the upper weight bounds
+  if (check_for_node(node, "upper_ww")) {
+    settings.upper_ww = get_node_array<double>(node, "upper_ww");  
+    // ensure there are enough entries
+    if (settings.upper_ww.size() != 
+      mesh_->n_bins()*(settings.energy_bounds.size()-1)) {
+         fatal_error("not enough entries in the upper_ww section");
+    }
+  }
+
+  // maybe multply out a vector for the upper bounds for consistency
+
+  return settings;
```

Comment:
I don't think you're going to get copy elision here, although I may be wrong. You should find a way to initialize this `settings` variable in-place rather than copying like this.